File: android/digia-engage/src/main/java/com/digia/engage/internal/VariableSchema.kt
package com.digia.engage.internal

internal data class VariableSchema(
    val name: String,
    val type: String,        // "string" | "number"
    val fallbackValue: String,
)

/** Normalises a raw JSON entry into a [VariableSchema]; null if it has no name. */
internal fun normalizeVariable(entry: org.json.JSONObject): VariableSchema? {
    val name = entry.optString("name").ifBlank { return null }
    val type = entry.optString("type").ifBlank { "string" }
    val fallbackValue = entry.optString("fallbackValue").ifBlank {
        entry.optString("sampleValue", "")
    }
    return VariableSchema(name = name, type = type, fallbackValue = fallbackValue)
}

File: flutter/lib/api/models/variable_schema.dart
/// Authoring metadata for one `{{ name }}` placeholder declared in the
/// Digia dashboard's `templateConfig.variables` list.
///
/// [type] determines how the value is handled at render time:
///   - `"string"` — substituted as-is.
///   - `"number"` — numeric coercion is attempted; participates in
///     arithmetic expressions inside `{{ }}`.
///
/// [fallbackValue] is the dashboard-declared default. Absent
/// `fallbackValue` falls back to `sampleValue` for backward-compatibility.
class VariableSchema {
  final String name;
  final String type; // 'string' | 'number'
  final String fallbackValue;

  const VariableSchema({
    required this.name,
    required this.type,
    required this.fallbackValue,
  });
}

/// Parses one raw entry from `templateConfig.variables`.
/// Returns `null` when the entry lacks a valid name.
VariableSchema? normalizeVariable(Map<String, dynamic> entry) {
  final name = entry['name'] as String?;
  if (name == null || name.isEmpty) return null;
  final type = entry['type'] == 'number' ? 'number' : 'string';
  final fallback =
      (entry['fallbackValue'] as String?) ?? (entry['sampleValue'] as String?) ?? '';
  return VariableSchema(name: name, type: type, fallbackValue: fallback);
}
